test-loop-tests/src/tests/catching_up.rs

```diff
@@ -0,0 +1,515 @@
+/// third epoch, and then making sure that the balances are correct for the next three epochs.
+/// If random one parts fetched during the epoch preceding the epoch a block producer is
+/// assigned to were to have incorrect receipts, the balances in the fourth epoch would have
+/// been incorrect due to wrong receipts applied during the third epoch.
+#[test]
+fn ultra_slow_test_catchup_random_single_part_sync() {
+    test_catchup_random_single_part_sync_common(false, false, 22)
+}
+
+// Same test as `test_catchup_random_single_part_sync`, but skips the chunks on height 23 and 24
+// It causes all the receipts to be applied only on height 25, which is the next epoch.
+// It tests that the incoming receipts are property synced through epochs
+#[test]
+fn ultra_slow_test_catchup_random_single_part_sync_skip_24() {
+    test_catchup_random_single_part_sync_common(true, false, 22)
+}
+
+#[test]
+fn ultra_slow_test_catchup_random_single_part_sync_send_24() {
+    test_catchup_random_single_part_sync_common(false, false, 24)
+}
+
+// Make sure that transactions are at least applied.
+#[test]
+fn ultra_slow_test_catchup_random_single_part_sync_non_zero_amounts() {
+    test_catchup_random_single_part_sync_common(false, true, 22)
+}
+
+// Use another height to send txs.
+#[test]
+fn ultra_slow_test_catchup_random_single_part_sync_height_9() {
+    test_catchup_random_single_part_sync_common(false, false, 9)
+}
+
+fn test_catchup_random_single_part_sync_common(skip_24: bool, non_zero: bool, send_tx_height: u64) {
+    init_test_logger();
+
+    let validators: Vec<Vec<AccountId>> = [
+        vec!["test1.1", "test1.2", "test1.3", "test1.4"],
+        vec!["test2.1", "test2.2", "test2.3", "test2.4"],
+        vec![
+            "test3.1", "test3.2", "test3.3", "test3.4", "test3.5", "test3.6", "test3.7", "test3.8",
+        ],
+    ]
+    .iter()
+    .map(|vec| vec.iter().map(|account| account.parse().unwrap()).collect())
+    .collect();
+    let seats: NumSeats = 8;
+
+    let stake = ONE_NEAR;
+    let mut runner = RotatingValidatorsRunner::new(stake, validators);
+
+    let test_accounts: Vec<AccountId> =
+        (0..16).map(|i| format!("test_account{i}").parse().unwrap()).collect_vec();
+
+    let all_accounts =
+        runner.all_validators_accounts().iter().chain(test_accounts.iter()).cloned().collect_vec();
+
+    let num_shards = 4;
+    let epoch_length = 8;
+    let genesis = TestLoopBuilder::new_genesis_builder()
+        .epoch_length(epoch_length)
+        .validators_spec(runner.genesis_validators_spec(seats, seats, seats))
+        .add_user_accounts_simple(&all_accounts, stake)
+        .shard_layout(ShardLayout::multi_shard(num_shards, 3))
+        .build();
+
+    let epoch_config_store = TestEpochConfigBuilder::from_genesis(&genesis)
+        .minimum_validators_per_shard(2)
+        .build_store_for_genesis_protocol_version();
+
+    let block_prod_time = Duration::milliseconds(6000);
```

Comment:
Is there a reason to set it to 6s? If not I would suggest using some default or 1s.

Reply:
Good point. I mostly kept it as it was before. Changed to default in [`43665f1` (#13419)](https://github.com/near/nearcore/pull/13419/commits/43665f1a769c8b9ed65e5cfa996d069074dd5954)

Reply:
Likely it was at 6 seconds to allow enough time for state sync when validators were changing.